source/beamline/BioXAS/BioXASCarbonFilterFarmActuatorControl.cpp

```diff
@@ -0,0 +1,333 @@
+
+	cancelledMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(cancelled()), cancelledMapper_, SLOT(map()) );
+
+	failedMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(failed()), failedMapper_, SLOT(map()) );
+
+	succeededMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(succeeded()), succeededMapper_, SLOT(map()) );
+
+	// Run action.
+
+	moveAction->start();
+
+	return AMControl::NoFailure;
+}
+
+void BioXASCarbonFilterFarmActuatorControl::setWindowPosition(Window::Selection window, double position)
+{
+	positionMap_.insert(window, position);
+}
+
+void BioXASCarbonFilterFarmActuatorControl::setWindow(double newWindow)
+{
+	if (value_ != newWindow) {
+		value_ = newWindow;
+		emit valueChanged(value_);
+	}
+}
+
+void BioXASCarbonFilterFarmActuatorControl::setWindowSetpoint(double newWindow)
+{
+	if (setpoint_ != newWindow) {
+		setpoint_ = newWindow;
+		emit setpointChanged(setpoint_);
+	}
+}
+
+void BioXASCarbonFilterFarmActuatorControl::setMoveInProgress(bool isMoving)
+{
+	if (moveInProgress_ != isMoving) {
+		moveInProgress_ = isMoving;
+		emit movingChanged(moveInProgress_);
+
+		if (moveInProgress_)
+			emit moveStarted();
+	}
+}
+
+void BioXASCarbonFilterFarmActuatorControl::onMoveStarted()
+{
+	setMoveInProgress(true);
+}
+
+void BioXASCarbonFilterFarmActuatorControl::onMoveCancelled(QObject *action)
+{
+	moveCleanup(action);
+	emit moveFailed(AMControl::WasStoppedFailure);
+}
+
+void BioXASCarbonFilterFarmActuatorControl::onMoveFailed(QObject *action)
+{
+	moveCleanup(action);
+	emit moveFailed(AMControl::OtherFailure);
+}
+
+void BioXASCarbonFilterFarmActuatorControl::onMoveSucceeded(QObject *action)
+{
+	moveCleanup(action);
+	emit moveSucceeded();
+}
+
+void BioXASCarbonFilterFarmActuatorControl::updateWindow()
+{
+	if (isConnected()) {
+		setWindow( windowAtPosition(position_->value()) );
+
+	} else {
+		setWindow(Window::None);
+	}
+}
+
+AMAction3* BioXASCarbonFilterFarmActuatorControl::createMoveAction(double setpoint)
+{
+	AMListAction3 *action = 0;
+
+	bool setpointOkay = validWindowSetpoint(setpoint);
+
+	if (isConnected() && setpointOkay) {
+		Window::Selection windowSetpoint = window(setpoint);
+		bool mappingOkay = positionMap_.contains(windowSetpoint);
+
+		if (mappingOkay) {
+			action = new AMListAction3(new AMListActionInfo3("Move BioXAS Carbon Filter Farm Actuator", "Move BioXAS Carbon Filter Farm Actuator"), AMListAction3::Sequential);
+
+			AMAction3 *move = AMActionSupport::buildControlMoveAction(position_, positionOfWindow(windowSetpoint));
+			action->addSubAction(move);
+
+			AMAction3 *check = AMActionSupport::buildControlWaitAction(status_, Position::Safe, TIMEOUT_MOVE);
+			action->addSubAction(check);
+		}
+	}
+
+	return action;
+}
+
+void BioXASCarbonFilterFarmActuatorControl::moveCleanup(QObject *action)
+{
+	if (action) {
+		disconnect( action, 0, cancelledMapper_, 0 );
```

Comment:
You could also just call ```action->disconnect();```.  That will disconnect the object from everything in a single function call.